Hackathons List Page

walidmujahid opened the pull request:
**Demo Video** : https://youtu.be/pYi6Ul-Uvc0
**Updsted Design - Copied checklist:** https://github.com/gitcoinco/web/issues/7096#issuecomment-661654141

* [x]  make the top header span across the page
* [x]  Contact Sales section  that was on the right side, is now below and CTA is Become a Sponsor linking to the same form.
* [x]  Remove Tribes section
* [x]  Remove Workshops  
* [x]  Update overall buttons so there is a primary and secondary focus
* [x]  on  the  hackathon card, add space for a few lines of text that gitcoin admin can enter in manually  using an  Edit button.
* [x]  add small sponsor logos to each hackathon

Happening Now tab:
![happening](https://user-images.githubusercontent.com/19514207/87563960-68271b00-c6f2-11ea-9611-c503307875d7.jpg)
 
* [x]  When hackathon  is completed - show projects and showcase buttons
 
 Completed tab:
![completed](https://user-images.githubusercontent.com/19514207/87564058-868d1680-c6f2-11ea-9753-78e934a724fb.jpg) 

closes https://github.com/gitcoinco/web/issues/7096

---

<details>
<summary>previous version</summary>
**Updated Demo Video - Full page with recent changes to hackathon section - 2020-06-11: https://youtu.be/rSdZAFy8gAc**
Demo video - workshops section various device widths - 2020-06-09: https://youtu.be/WB7YmUwbBc4
Demo  video - Workshops - 2020-06-07: https://youtu.be/C2M6hholtZc
Demo of Progress from 2020-04-06 05:26: https://youtu.be/wKL6WvQ3Sbs 

Main change on front end is update to margins for the boxes in the header. 

Currently:

The boxes in the header can feature hackathons and the process to feature is shown in the video. When there is no featured hackathon, the header looksl like this:

![123497](https://user-images.githubusercontent.com/1489224/79135928-1f239b80-7d7e-11ea-806f-5c3101037ed8.png)

The next section is the list of hackathons categorised as 'current', 'upcoming', and 'finished'. Each category can be navigated via the respective tabs. The 'current' tab is the default tab and if there are no 'current' hackathons then the 'upcoming' tab is the default if there are any, otherwise, the 'finished' tab is the default.

Next is the Tribes section where it shows all tribes that have posted prizes for all current hackathons. If there are no current hackathons but there exists upcoming hackathons, then tribes who had posted prizes in the most recently active hackathon are shown. If there are neither current nor upcoming hackathons, then the top four tribes are shown in order of the number of hackathons the respective tribes participated in.

<strike>Workshops will be added in a separate PR.</strike> Workshops added.

---

Need to:

- [ ] refactor code

https://github.com/gitcoinco/web/issues/6211
</details>

walidmujahid commented:
@PixelantDesign @androolloyd

androolloyd commented:
@walidmujahid You'll have to create your migrations. They are failing as is in Travis. 

If you have any issues, message me on Gitcoin Chat and I can help.

walidmujahid commented:
@androolloyd Okay, the migration problem is fixed.

walidmujahid commented:
I will update this with most recent master and recreate migrations.

Has this been reveiwed @androolloyd? What changes are there that I need to make?

androolloyd commented:
I will have a look @walidmujahid

walidmujahid commented:
This has been updated with the most recent changes in master and migrations recreated. 

@PixelantDesign @danlipert @thelostone-mc @octavioamu

walidmujahid commented:
@PixelantDesign I have added the workshops template markup.

![workshops](https://user-images.githubusercontent.com/1489224/82017265-bbc4ac00-9650-11ea-91f7-cb7df9767977.png)

I currently have the following fields in the model:

https://github.com/walidmujahid/web/blob/f667d857a42c678fd19516cfe82f613743d6d963/app/dashboard/models.py#L4967-L4985

Admin view:

![admin-workshops](https://user-images.githubusercontent.com/1489224/82017428-1fe77000-9651-11ea-87f3-3497e5ad9689.png)

I still need to add the speaker logo.

thelostone-mc commented:
@walidmujahid I've shifted this to draft ! Once you've added stuff and it's got for review -> mark it as ready for review ^_^

fluffays commented:
Adding suggested copy for the header boxes - @PixelantDesign 

_For the main box:_

**Hack for the best open source projects,
earn prizes & find your OSS tribe**

Join Gitcoin’s hackathons and help create better solutions for the future. Find your dream team among our 36k hackers and compete for solid crypto rewards. OSS wants you!

**CTA**:  CHECK HACKATHONS

----------------------------------------------------------
_For the “get-in-touch” section:_

Want to host your own hackathon?

**CTA**: LET’S TALK

PixelantDesign commented:
@walidmujahid would you mind updating the copy per suggestion from @fluffays ?

walidmujahid commented:
> @walidmujahid would you mind updating the copy per suggestion from @fluffays ?

I just saw the messages now @PixelantDesign. I will update quickly.

walidmujahid commented:
Workshops section has also been updated in the most recent push @PixelantDesign. It looks like this:

![workshops3](https://user-images.githubusercontent.com/1489224/82466183-136a7980-9a8e-11ea-934b-a872caf71a89.png)

walidmujahid commented:
> Adding suggested copy for the header boxes - @PixelantDesign
> 
> _For the main box:_
> 
> **Hack for the best open source projects, earn prizes & find your OSS tribe**
> 
> Join Gitcoin’s hackathons and help create better solutions for the future. Find your dream team among our 36k hackers and compete for solid crypto rewards. OSS wants you!
> 
> **CTA**: CHECK HACKATHONS
> 
> _For the “get-in-touch” section:_
> 
> Want to host your own hackathon?
> 
> **CTA**: LET’S TALK

Pardon @fluffays @PixelantDesign, but by 'main box' do you mean this:

![non-featured](https://user-images.githubusercontent.com/1489224/82466664-b3c09e00-9a8e-11ea-8e20-c3e050b217eb.png)

And what do you mean by CTA?

fluffays commented:
> Pardon @fluffays @PixelantDesign, but by 'main box' do you mean this:
> 
> ![non-featured](https://user-images.githubusercontent.com/1489224/82466664-b3c09e00-9a8e-11ea-8e20-c3e050b217eb.png)
> 
> And what do you mean by CTA?

Heys @walidmujahid !

Yes, "main box" = the one in your screenshot. By CTA I mean adding a button with the respective text - "CHECK HACKATHONS" for the main box, "LET'S TALK" for the other box.

Thanks! 🙏

walidmujahid commented:
> > Pardon @fluffays @PixelantDesign, but by 'main box' do you mean this:
> > ![non-featured](https://user-images.githubusercontent.com/1489224/82466664-b3c09e00-9a8e-11ea-8e20-c3e050b217eb.png)
> > And what do you mean by CTA?
> 
> Heys @walidmujahid !
> 
> Yes, "main box" = the one in your screenshot. By CTA I mean adding a button with the respective text - "CHECK HACKATHONS" for the main box, "LET'S TALK" for the other box.
> 
> Thanks! 🙏

Alright @fluffays @PixelantDesign, I have updated the copy.

![updated-copy](https://user-images.githubusercontent.com/1489224/82560673-e24a8180-9b3f-11ea-9bde-9398d4ada299.png)

Just to note, the main box does not have a button, however, should does it need a button for the mentioned 'CTA'?

walidmujahid commented:
> Looks good once @walidmujahid could you move open internet image up a little bit?

Is this better @PixelantDesign?

![reposition-copy-padding](https://user-images.githubusercontent.com/1489224/82579943-82160880-9b5c-11ea-8a65-330e12bf211a.png)

walidmujahid commented:
@PixelantDesign @androolloyd @danlipert @thelostone-mc @octavioamu This PR is functional. Is there anything I need to do to help speed this along?

walidmujahid commented:
@PixelantDesign @androolloyd @danlipert @thelostone-mc @octavioamu This PR is functional. Is there anything I need to do to help speed merging into master?

walidmujahid commented:
@PixelantDesign 

Merging with current upstream changes, specifically this pull request: https://github.com/gitcoinco/web/pull/6722

The following changes have been made to the hackathon cards, so they look like this now:

![updated-cards](https://user-images.githubusercontent.com/1489224/83319610-1263e580-a20e-11ea-8a18-cf49183336b2.png)

PixelantDesign commented:
Thanks @walidmujahid. Sorry for the confusion.

@thelostone-mc how do we get the rest of @walidmujahid's pr in yours? The thumbnails are particularly small and we're missing the top section with the contact sales info as well as blog calendaring section at the bottom. Also @frankchen07 @vs77bb and I weren't aware you were building on top of this pr.

thelostone-mc commented:
@walidmujahid Thumbail being and stuff is not something you'd need to worry about that !
Those were my changes and I've got a PR https://github.com/gitcoinco/web/pull/6766 to fix those up. Ideally they won't affect what you've written up here

walidmujahid commented:
> @walidmujahid Thumbail being and stuff is not something you'd need to worry about that !
> Those were my changes and I've got a PR #6766 to fix those up. Ideally they won't affect what you've written up here

@thelostone-mc I will update this branch with master a bit later. As for the PR #6766, is it scheduled to be merged into master soon? Or if not, is it safe to merge into my PR?

octavioamu commented:
hey @walidmujahid  I don't see anything related to the workshop part on the video, that is blocking the feature review from the product team. Can you show it working?

walidmujahid commented:
> hey @walidmujahid I don't see anything related to the workshop part on the video, that is blocking the feature review from the product team. Can you show it working?

Right. sorry, I forgot to upload a new demo video @octavioamu

walidmujahid commented:
@PixelantDesign @thelostone-mc After merging #6722 and #6766 updates, the following is how the listing section appears -includes updates from my branch in commits 207c6b32e3b915d252bf53540505fe9fd92d5464 and 2796ee343ec52f23de097df6f01d4dbbb3813a4f:

![updates-to-hackathon-section](https://user-images.githubusercontent.com/1489224/83970943-7fc6e480-a8a6-11ea-9e98-fb26727bac97.png)

walidmujahid commented:
> hey @walidmujahid I don't see anything related to the workshop part on the video, that is blocking the feature review from the product team. Can you show it working?

@octavioamu @PixelantDesign I have uploaded a demo video for the workshop section: https://youtu.be/C2M6hholtZc

octavioamu commented on app/dashboard/admin.py:
```diff
@@ -327,6 +330,12 @@ class HackathonSponsorAdmin(admin.ModelAdmin):
+class HackathonWorkshopAdmin(admin.ModelAdmin):
+    """The admin object for the HackathonWorkshop model."""
+
+    list_display = ['pk', 'start_date', 'hackathon', 'speaker', 'url']
```
let's change here speaker to raw_id_fields.

walidmujahid replied:
Fixed @octavioamu

octavioamu commented on app/dashboard/templates/dashboard/hackathon/hackathons.html:
```diff
@@ -77,79 +167,183 @@
+                          <div class="row">
+                            <div>
+                              <a href="{{ workshop.url }}" target="_blank" style="text-decoration:none;">
+                                <img id="workshop-cover" src="{{ MEDIA_URL }}{{ workshop.cover }}" alt="Workshop cover" />
```
this should be a class as is inside a loop and you cannot have more than one ID with the same name.

octavioamu replied:
also probably this is not needed, adding mw-100 (max-width: 100% bootstrap class)  will handle to avoid the image go further the column. 
Of course people using images need to be attached to a ratio to avoid some of the images being taller than the others.

walidmujahid replied:
@octavioamu removed in 2398f7af31

frankchen07 commented:
the projects page looks like it has a new design, but it seems like the elements are misaligned, and the project cards look a little "rough". Not sure if that's in the scope of this ticket though. @PixelantDesign did we have some kind of design we were working off of where we can give comments?

what does it actually look like when someone clicks on the workshop after creating it?

I would think there should be some kind of zoom link too, but perhaps I'm mistaken or it's out of scope. 

something to think about in a v2, some kind of calendar integration where people can add to their own calendars.

octavioamu commented:
@frankchen07  I believe that part of the cards come from another PR already merged so not really part of this one.

octavioamu commented:
> what does it actually look like when someone clicks on the workshop after creating it?

Is a link field so it goes to what ever link we put on the admin, can be a zoom one or another thing.

walidmujahid commented:
> @frankchen07 I believe that part of the cards come from another PR already merged so not really part of this one.

Yes. Specifically #6722 #6766

walidmujahid commented:
@octavioamu I have updated to include the fixes you requested among other things in: 6094e9a and 2398f7a

walidmujahid commented:
> @octavioamu I have updated to include the fixes you requested among other things in: [6094e9a](https://github.com/gitcoinco/web/commit/6094e9a0998fe7285409efbe257421f776d10af3) and [2398f7a](https://github.com/gitcoinco/web/commit/2398f7af318e472b5b13e6c45e9605bfea32480f)

@PixelantDesign I have uploaded a second for the changes that show the workshop section for various device widths: https://youtu.be/WB7YmUwbBc4

Widths tested: 

above 1200px, between 1024px and ~900px, between ~900px and 768px, and beneath 768px.

walidmujahid commented:
> Looking good! Is it possible to get a video of the entire page?

Sorry for the late upload. Here is a video of the full page @PixelantDesign: https://youtu.be/rSdZAFy8gAc

 I added only one bounty to one hackathon, so only one tribe is listed.

walidmujahid commented:
Sorry, I probably should have pinged people again. This is still open for review. @thelostone-mc @octavioamu

PixelantDesign commented:
Hi @walidmujahid would you mind posting a video or this pr as it stands?

walidmujahid commented:
> Hi @walidmujahid would you mind posting a video or this pr as it stands?

@PixelantDesign Sure. I had shared the video last week and nothing has changed on the frontend since uploading: https://youtu.be/rSdZAFy8gAc

As for this PR, I am waiting for any further review and feedback. I will need to resolve conflicts for recent changes in upstream master.

frankchen07 commented:
I think current cards on `hackathon-list` seem to be updated? They look slightly different.

Is the idea that the current running hackathons will be displayed prominently? In the video above, it's the COVID hackathon. What happens if there are 2-3 concurrent hackathons?

walidmujahid commented:
> I think current cards on `hackathon-list` seem to be updated? They look slightly different.

Yes. The changes were introduced in PRs: #6722 and #6766 

> Is the idea that the current running hackathons will be displayed prominently? In the video above, it's the COVID hackathon. What happens if there are 2-3 concurrent hackathons?

I think that is the idea. @PixelantDesign Would probably have the answer for these questions. Changing the featured hackathon can be done in the select hackathon event in admin: https://youtu.be/wKL6WvQ3Sbs?t=127

walidmujahid commented:
> I think current cards on `hackathon-list` seem to be updated? They look slightly different.

@frankchen07 Yes. The changes were introduced in PRs: #6722 and #6766 

> Is the idea that the current running hackathons will be displayed prominently? In the video above, it's the COVID hackathon. What happens if there are 2-3 concurrent hackathons?

I think that is the idea. @PixelantDesign Would probably have the answer for these questions. Changing the featured hackathon can be done in the select hackathon event in admin: https://youtu.be/wKL6WvQ3Sbs?t=127

octavioamu commented on app/dashboard/templates/dashboard/hackathon/hackathons.html:
```diff
@@ -77,79 +181,178 @@
+                {% for workshop in workshops %}
+                    <div class="col-xs-12 col-sm-7 col-md-7 col-lg-5 col-xl-4">
+                      <div class="mb-5 mt-5">
+                        <div style="position:relative">
```
bootstrap class can be used here `position-relative`

walidmujahid replied:
@octavioamu refactored to use the bootstrap class.

octavioamu commented on app/dashboard/templates/dashboard/hackathon/hackathons.html:
```diff
@@ -77,79 +181,178 @@
     {% include 'shared/footer_scripts.html' %}
     {% include 'shared/footer.html' %}
 
+    {% block 'scripts' %}
+    <!-- <script src="{% static "v2/js/lib/jquery-unveil.js" %}"></script> -->
+    <script src="{% static "v2/js/pages/hackathon-list.js" %}"></script>
+    <!-- <script>
+      $(document).ready(function() {
+        setInterval(function(){
+          var scrollTop = window.pageYOffset || document.documentElement.scrollTop;
+          if (scrollTop > 500 && !document.offset_loaded){
+            $("img").unveil(200);
+            document.offset_loaded = true;
+          }
+        }, 3000);
+        setTimeout(function(){
+          $('.leaderboard_hero').css('background-image', 'url("{% static 'v2/images/kudos/kudos-bg.png' %}")');
+        }, 1000);
+      });
+    </script> -->
+    {% endblock %}
```
dead code

walidmujahid replied:
@octavioamu removed the lines

octavioamu commented:
@walidmujahid when this will be ready? lets focus on get it finished to deploy it

walidmujahid commented:
@octavioamu I will update soon as per the commented feedback.

> @walidmujahid when this will be ready? lets focus on get it finished to deploy it

@octavioamu I have been informed recently that some additions had been discussed, so @PixelantDesign is working on some design mockups for them. I am not sure how you guys might want to handle this? Perhaps wait until I recieve the mockups and push the changes? Or perhaps merge this PR into master and open a new PR for the additions?

walidmujahid commented:
This PR has recently benn update to include the changed design from https://github.com/gitcoinco/web/issues/7096

Demo Video: https://youtu.be/pYi6Ul-Uvc0
I forgot to show it in the video, however, when clicking on the small sponsor avatar in a hackathon card, the sponsor's profile is opened in a new tab.

@octavioamu @PixelantDesign @thelostone-mc @danlipert @frankchen07

danlipert commented on app/dashboard/models.py:
```diff
@@ -4791,7 +4794,8 @@ def psave_hackathonevent(sender, instance, **kwargs):
                 "visible_to":None,
                 'img_url': instance.logo.url if instance.logo else None,
             }
-            )
```
please dont make arbitrary style changes like this as it makes your PRs much harder to review

walidmujahid replied:
I am sure when the change was introduced. I shall revert it.

walidmujahid replied:
@danlipert I am sure when the change was introduced. I shall revert it.

danlipert commented on app/dashboard/templatetags/math_filters.py:
```diff
@@ -0,0 +1,14 @@
+import re
```
what is this file for?

walidmujahid replied:
It is no longer relevant. Will delete @danlipert

thelostone-mc commented on app/assets/v2/css/home.css:
```diff
@@ -510,6 +510,13 @@ figure:hover::before {
+  .box-cta {
+    margin: 0 !important;
+  }
+  .work-internet {
```
^ add empty line

thelostone-mc commented on app/assets/v2/js/pages/hackathon-list.js:
```diff
@@ -0,0 +1,20 @@
+$(document).ready(function() {
+  $(document).on('click', '#tabs a', function(e) {
+    e.preventDefault();
+    var target = $(this).data('href');
```
use const / let

walidmujahid commented:
Pulled most recent upstream changes and resolved conflicts.

@danlipert @thelostone-mc 
I have made fixes based on feedback:
- removed unused files.
- reverted to previous spacing in models hackathon presave signal to avoid unnessecary style change.
- added empty line at L517 in `home.css`.
- using `let` instead of `var` in `hackathon-list.js`

walidmujahid commented:
@danlipert @thelostone-mc @octavioamu What is needed to get this merged?

thelostone-mc commented on app/dashboard/utils.py:
```diff
@@ -1087,3 +1087,18 @@ def get_token_recipient_senders(network, recipient_address, token_address):
     return [process_log(log) for log in logs]
+
+
+def get_hackathon_event(title, event, network):
```
@owocki  It's this which is eats up the loadtime from the looks of it ! 
Cause we invoke this tice

owocki replied:
makes sense. i think persisting this info in the model will solve muhc of the issues

owocki replied:
and i defer to the engineers but i think a limit on the # of db calls per pageload is a good idea, as a code standsrad

walidmujahid replied:
> makes sense. i think persisting this info in the model will solve muhc of the issues

I would like to help resolve this @thelostone-mc @owocki. However, I have yet to work with the presistance methods in Gitcoin project. With that said, are you basically looking for the `event_bounties` and `registrants` in the `get_hackathon_event(title, event, network)` function to be cached using the `.cache()` method from the installed django-cacheops?

walidmujahid replied:
> makes sense. i think persisting this info in the model will solve muhc of the issues

I would like to help resolve this @thelostone-mc @owocki. However, I have yet to work with the presistance methods in Gitcoin project. With that said, are you basically looking for the `event_bounties` and `registrants` querysets in the `get_hackathon_event(title, event, network)` function to be cached using the `.cache()` method from the installed django-cacheops?

walidmujahid replied:
> makes sense. i think persisting this info in the model will solve muhc of the issues

I would like to help resolve this @thelostone-mc @owocki. However, I have yet to work with the presistance methods in Gitcoin project. With that said, are you basically looking for the `event_bounties` and `registrants` querysets in the `get_hackathon_event(title, event, network)` function to be cached using the `.cache()` method from the installed django-cacheops? Or perhaps caching the `get_hackathon_event` function?

owocki replied:
i'd just throw them in a new JSONField on the object, and then make sure the `create_page_cache` mgmt command refreshes them whenever its fun

the `.cache()` thing only works if you have something that periodically warms the cache, so its not ideal

owocki replied:
@walidmujahid any update? the page is still loading in over 2 seconds and will get worse with every hack we add

walidmujahid replied:
@thelostone-mc @owocki I had missed the fact that the `event_bounties` and `registrants` have now become obsolete and are currently unused. I have removed them -draft PR https://github.com/gitcoinco/web/pull/7254 -, however, I am not sure if this helps resolve the main loadtime issue seen as I have yet to locally test this. 

On the matter of persisting the conetxt or certain information for the context, I believed I had the solution, however, I am a bit uncertain about it. The information in the `get_hackathon_event` util function includes a badly named  key called `title` -which helps categorise the hackathon event based on whether it is current, upcoming, or finished- and `hackathon` which is just a querystring. 

What I am hesitant about is putting `title` in a JSONField in the object, and then refreshing whether it is `current`, or `finished`, or otherwise because, if I do, would the cron delay for running `create_page_cache` cause some confusion? Is the cron delay something to be concerned about in the first place? If yes then would putting all finished hackathons in an `JSONStore` in `create_page_cache` help solve the laodtime issue? And if the cron delay is not much of an issue, the would it make sense to just go ahead and have `create_page_cache` store the entire context in a `JSONStore` and as such newly added hackathons or updates will appear whenever the cron runs?

If I am overcomplicating things, forgive me and let me know.

owocki replied:
i am willing to settle for a slight delay in updating the hack info if it means the page loads in 0.5s instead of 2.5s.  i say proceed with the caching approach!

pls invoice me gitcoin.co/requestmoney when u have a PR out

walidmujahid replied:
Since PR https://github.com/gitcoinco/web/pull/7254 has been merge I am going to set this conversation as resolved.
